File: packages/aid-go/pka.go
package aid

import (
    "crypto/ed25519"
    "encoding/base64"
    "fmt"
    "net/http"
    "net/url"
    "strings"
    "time"
)

// performPKAHandshake performs an RFC 9421 Ed25519 verification against the agent endpoint.
func performPKAHandshake(uri, pka, kid string, timeout time.Duration) error {
    if kid == "" {
        return newAidError("ERR_SECURITY", "Missing kid for PKA")
    }
    u, err := url.Parse(uri)
    if err != nil || u.Host == "" {
        return newAidError("ERR_SECURITY", "Invalid URI for handshake")
    }
    // Prepare GET with challenge and date
    nonce := make([]byte, 32)
    // time-based pseudo randomness is sufficient for challenge uniqueness here
    for i := range nonce { nonce[i] = byte(time.Now().UnixNano() >> (i%8)) }
    challenge := base64.RawURLEncoding.EncodeToString(nonce)
    date := time.Now().UTC().Format("Mon, 02 Jan 2006 15:04:05 GMT")

    req, _ := http.NewRequest("GET", uri, nil)
    req.Header.Set("AID-Challenge", challenge)
    req.Header.Set("Date", date)
    client := *httpClient
    client.Timeout = timeout
    // Do not follow redirects for handshake per security policy
    client.CheckRedirect = func(req *http.Request, via []*http.Request) error { return http.ErrUseLastResponse }
    resp, err := client.Do(req)
    if err != nil {
        return newAidError("ERR_SECURITY", err.Error())
    }
    defer resp.Body.Close()
    if resp.StatusCode != 200 {
        return newAidError("ERR_SECURITY", fmt.Sprintf("Handshake HTTP %d", resp.StatusCode))
    }

    sigInput := resp.Header.Get("Signature-Input")
    if sigInput == "" { sigInput = resp.Header.Get("signature-input") }
    sig := resp.Header.Get("Signature")
    if sig == "" { sig = resp.Header.Get("signature") }
    if sigInput == "" || sig == "" {
        return newAidError("ERR_SECURITY", "Missing signature headers")
    }

    covered, created, keyidRaw, alg, signature, perr := parseSignatureHeaders(sigInput, sig)
    if perr != nil {
        return perr
    }
    now := time.Now().Unix()
    if created < now-300 || created > now+300 {
        return newAidError("ERR_SECURITY", "Signature created timestamp outside acceptance window")
    }
    // strip optional quotes around keyid for comparison
    keyid := keyidRaw
    if len(keyid) >= 2 && keyid[0] == '"' && keyid[len(keyid)-1] == '"' {
        keyid = keyid[1:len(keyid)-1]
    }
    if keyid != kid {
        return newAidError("ERR_SECURITY", "Signature keyid mismatch")
    }
    if strings.ToLower(alg) != "ed25519" {
        return newAidError("ERR_SECURITY", "Unsupported signature algorithm")
    }

    dateHeader := resp.Header.Get("Date")
    // Validate Date header if present (±300s window)
    if dateHeader != "" {
        if t, e := http.ParseTime(dateHeader); e == nil {
            now := time.Now().UTC()
            diff := t.Sub(now)
            if diff < 0 { diff = -diff }
            if diff > 300*time.Second {
                return newAidError("ERR_SECURITY", "HTTP Date header outside acceptance window")
            }
        } else {
            return newAidError("ERR_SECURITY", "Invalid Date header")
        }
    }

    base, berr := buildSignatureBase(covered, created, keyidRaw, alg, "GET", uri, u.Host, chooseDate(date, dateHeader), challenge)
    if berr != nil {
        return berr
    }
    pub, derr := multibaseDecode(pka)
    if derr != nil {
        return derr
    }
    if len(pub) != ed25519.PublicKeySize {
        return newAidError("ERR_SECURITY", "Invalid PKA length")
    }
    if !ed25519.Verify(ed25519.PublicKey(pub), base, signature) {
        return newAidError("ERR_SECURITY", "PKA signature verification failed")
    }
    return nil
}

func parseSignatureHeaders(sigInput, sig string) (covered []string, created int64, keyidRaw, alg string, signature []byte, err error) {
    // sig=("a" "b");created=...;keyid=...;alg="ed25519"
    idx := strings.Index(sigInput, "sig=(")
    if idx < 0 { return nil, 0, "", "", nil, newAidError("ERR_SECURITY", "Invalid Signature-Input") }
    rest := sigInput[idx+5:]
    close := strings.Index(rest, ")")
    if close < 0 { return nil, 0, "", "", nil, newAidError("ERR_SECURITY", "Invalid Signature-Input") }
    inside := rest[:close]
    // Extract quoted tokens
    for len(inside) > 0 {
        start := strings.Index(inside, "\"")
        if start < 0 { break }
        inside = inside[start+1:]
        end := strings.Index(inside, "\"")
        if end < 0 { break }
        covered = append(covered, inside[:end])
        inside = inside[end+1:]
    }
    if len(covered) == 0 { return nil, 0, "", "", nil, newAidError("ERR_SECURITY", "Invalid Signature-Input") }
    // Enforce exact required set
    required := map[string]struct{}{"aid-challenge":{}, "@method":{}, "@target-uri":{}, "host":{}, "date":{}}
    lower := map[string]struct{}{}
    for _, c := range covered { lower[strings.ToLower(c)] = struct{}{} }
    if len(lower) != len(required) {
        return nil, 0, "", "", nil, newAidError("ERR_SECURITY", "Signature-Input must cover required fields")
    }
    for r := range required { if _, ok := lower[r]; !ok { return nil, 0, "", "", nil, newAidError("ERR_SECURITY", "Signature-Input must cover required fields") } }

    // Params
    for _, part := range strings.Split(sigInput, ";") {
        p := strings.TrimSpace(part)
        if strings.HasPrefix(strings.ToLower(p), "created=") {
            var c int64
            _, e := fmt.Sscanf(p[len("created="):], "%d", &c)
            if e == nil { created = c }
        } else if strings.HasPrefix(strings.ToLower(p), "keyid=") {
            keyidRaw = strings.TrimSpace(p[len("keyid="):])
        } else if strings.HasPrefix(strings.ToLower(p), "alg=") {
            alg = strings.Trim(strings.TrimSpace(p[len("alg="):]), "\"")
        }
    }
    if created == 0 || keyidRaw == "" || alg == "" {
        return nil, 0, "", "", nil, newAidError("ERR_SECURITY", "Invalid Signature-Input")
    }

    // Signature header: sig=:base64:
    s := sig
    i := strings.Index(strings.ToLower(s), "sig=")
    if i < 0 { return nil, 0, "", "", nil, newAidError("ERR_SECURITY", "Invalid Signature header") }
    s = s[i+4:]
    if !strings.HasPrefix(s, ":") { return nil, 0, "", "", nil, newAidError("ERR_SECURITY", "Invalid Signature header") }
    s = s[1:]
    end := strings.Index(s, ":")
    if end < 0 { return nil, 0, "", "", nil, newAidError("ERR_SECURITY", "Invalid Signature header") }
    val := s[:end]
    dec, e := base64.StdEncoding.DecodeString(val)
    if e != nil { return nil, 0, "", "", nil, newAidError("ERR_SECURITY", "Invalid Signature header") }
    return covered, created, keyidRaw, alg, dec, nil
}

func chooseDate(requestDate, responseDate string) string {
    if responseDate != "" { return responseDate }
    return requestDate
}

func buildSignatureBase(covered []string, created int64, keyid, alg, method, targetURI, host, date, challenge string) ([]byte, error) {
    lines := make([]string, 0, len(covered)+1)
    for _, item := range covered {
        switch strings.ToLower(item) {
        case "aid-challenge":
            lines = append(lines, "\"AID-Challenge\": "+challenge)
        case "@method":
            lines = append(lines, "\"@method\": "+method)
        case "@target-uri":
            lines = append(lines, "\"@target-uri\": "+targetURI)
        case "host":
            lines = append(lines, "\"host\": "+host)
        case "date":
            lines = append(lines, "\"date\": "+date)
        default:
            return nil, newAidError("ERR_SECURITY", "Unsupported covered field: "+item)
        }
    }
    quoted := make([]string, len(covered))
    for i, c := range covered { quoted[i] = "\""+c+"\"" }
    params := fmt.Sprintf("(%s);created=%d;keyid=%s;alg=\"%s\"", strings.Join(quoted, " "), created, keyid, strings.ToLower(alg))
    lines = append(lines, "\"@signature-params\": "+params)
    return []byte(strings.Join(lines, "\n")), nil
}

func multibaseDecode(s string) ([]byte, error) {
    if s == "" { return nil, newAidError("ERR_SECURITY", "Empty PKA") }
    if s[0] != 'z' { return nil, newAidError("ERR_SECURITY", "Unsupported multibase prefix") }
    out, err := base58Decode(s[1:])
    if err != nil { return nil, err }
    return out, nil
}

func base58Decode(s string) ([]byte, error) {
    const alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
    if s == "" { return []byte{}, nil }
    zeros := 0
    for zeros < len(s) && s[zeros] == '1' { zeros++ }
    size := (len(s)-zeros)*733/1000 + 1 // log(58)/log(256) ≈ 0.733
    b := make([]byte, size)
    for i := zeros; i < len(s); i++ {
        ch := s[i]
        idx := strings.IndexByte(alphabet, ch)
        if idx < 0 { return nil, newAidError("ERR_SECURITY", "Invalid base58 character") }
        carry := idx
        for j := size - 1; j >= 0; j-- {
            carry += 58 * int(b[j])
            b[j] = byte(carry & 0xff)
            carry >>= 8
        }
    }
    // strip leading zeros
    it := 0
    for it < len(b) && b[it] == 0 { it++ }
    out := make([]byte, zeros+len(b)-it)
    for i := 0; i < zeros; i++ { out[i] = 0 }
    copy(out[zeros:], b[it:])
    return out, nil
}

File: packages/aid-py/aid_py/pka.py
"""PKA handshake (Python) for AID v1.1

Performs an HTTP Message Signatures (RFC 9421) verification using Ed25519 when
the TXT record includes a Public Key for Agent (pka) and key id (kid).
"""
from __future__ import annotations

import base64
import os
import re
import time
import hmac # Added for constant-time comparisons
from urllib.parse import urlparse
import urllib.request
import urllib.error

from .parser import AidError
import pathlib
import logging # Added for logging in empty except block

def _debug_write(name: str, data: str) -> None:
    try:
        d = pathlib.Path(__file__).resolve().parent / "_debug"
        d.mkdir(exist_ok=True)
        (d / name).write_text(data)
    except Exception as e:
        # Intentionally swallowing error here for debug writing,
        # as failure to write debug data should not stop the main flow.
        # Log for visibility in debug builds.
        logging.debug(f"Failed to write debug data: {e}")


def _b58_decode(s: str) -> bytes:
    alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
    if not s:
        return b""
    zeros = 0
    while zeros < len(s) and s[zeros] == "1":
        zeros += 1
    # approximate size
    size = int(((len(s) - zeros) * (58).bit_length() / 8) + 1)
    b = [0] * size
    for ch in s[zeros:]:
        try:
            val = alphabet.index(ch)
        except ValueError:
            raise AidError("ERR_SECURITY", "Invalid base58 character") from None
        carry = val
        for j in range(size - 1, -1, -1):
            carry += 58 * b[j]
            b[j] = carry & 0xFF
            carry >>= 8
    # strip leading zeros in b
    it = 0
    while it < len(b) and b[it] == 0:
        it += 1
    out = bytes([0] * zeros + b[it:])
    return out


def _multibase_decode(s: str) -> bytes:
    if not s:
        raise AidError("ERR_SECURITY", "Empty PKA")
    prefix, payload = s[0], s[1:]
    if prefix == "z":
        return _b58_decode(payload)
    raise AidError("ERR_SECURITY", "Unsupported multibase prefix")


def _parse_signature_headers(headers: dict[str, str]) -> tuple[list[str], int, str, str, str, bytes, str | None]:
    sig_input = headers.get("Signature-Input") or headers.get("signature-input")
    sig = headers.get("Signature") or headers.get("signature")
    if not sig_input or not sig:
        raise AidError("ERR_SECURITY", "Missing signature headers")

    inside = re.search(r"sig=\(\s*([^)]*?)\s*\)", sig_input, flags=re.I)
    if not inside:
        raise AidError("ERR_SECURITY", "Invalid Signature-Input")
    covered: list[str] = re.findall(r'"([^"]+)"', inside.group(1))
    if not covered:
        raise AidError("ERR_SECURITY", "Invalid Signature-Input")
    required = {"aid-challenge", "@method", "@target-uri", "host", "date"}
    lowers = {c.lower() for c in covered}
    if lowers != required:
        raise AidError("ERR_SECURITY", "Signature-Input must cover required fields")

    cm = re.search(r"(?:^|;)\s*created=(\d+)", sig_input, flags=re.I)
    km = re.search(r"(?:^|;)\s*keyid=([^;\s]+)", sig_input, flags=re.I)
    am = re.search(r"(?:^|;)\s*alg=\"([^\"]+)\"", sig_input, flags=re.I)
    if not cm or not km or not am:
        raise AidError("ERR_SECURITY", "Invalid Signature-Input")
    created = int(cm.group(1))
    keyid_raw = km.group(1)
    keyid = keyid_raw.strip('"')
    alg = am.group(1).lower()

    sm = re.search(r"sig\s*=\s*:\s*([^:]+)\s*:", sig, flags=re.I)
    if not sm:
        raise AidError("ERR_SECURITY", "Invalid Signature header")
    signature = base64.b64decode(sm.group(1))
    date_header = headers.get("Date") or headers.get("date")
    return covered, created, keyid, keyid_raw, alg, signature, date_header


def _build_signature_base(
    covered: list[str],
    *,
    created: int,
    keyid: str,
    alg: str,
    method: str,
    target_uri: str,
    host: str,
    date: str,
    challenge: str,
) -> bytes:
    lines: list[str] = []
    for item in covered:
        lower = item.lower()
        if lower == "aid-challenge":
            lines.append(f'"AID-Challenge": {challenge}')
        elif lower == "@method":
            lines.append(f'"@method": {method}')
        elif lower == "@target-uri":
            lines.append(f'"@target-uri": {target_uri}')
        elif lower == "host":
            lines.append(f'"host": {host}')
        elif lower == "date":
            lines.append(f'"date": {date}')
        else:
            raise AidError("ERR_SECURITY", f"Unsupported covered field: {item}")
    quoted = " ".join(f'"{c}"' for c in covered)
    params = f"({quoted});created={created};keyid={keyid};alg=\"{alg}\""
    lines.append(f'"@signature-params": {params}')
    return "\n".join(lines).encode("utf-8")


def perform_pka_handshake(uri: str, pka: str, kid: str, *, timeout: float = 2.0) -> None:
    if not kid:
        raise AidError("ERR_SECURITY", "Missing kid for PKA")
    parsed = urlparse(uri)
    if not parsed.scheme or not parsed.netloc:
        raise AidError("ERR_SECURITY", "Invalid URI for handshake")

    # Prepare request
    nonce = os.urandom(32)
    challenge = base64.urlsafe_b64encode(nonce).decode("ascii").rstrip("=")
    date_hdr = time.strftime("%a, %d %b %Y %H:%M:%S GMT", time.gmtime())
    req = urllib.request.Request(uri, headers={"AID-Challenge": challenge, "Date": date_hdr})

    try:
        opener = urllib.request.build_opener()
        with opener.open(req, timeout=timeout) as resp:  # nosec B310
            if resp.status != 200:
                raise AidError("ERR_SECURITY", f"Handshake HTTP {resp.status}")
            headers = {k: v for k, v in resp.headers.items()}
    except AidError:
        raise
    except Exception as exc:  # pragma: no cover - network errors
        raise AidError("ERR_SECURITY", str(exc)) from None

    covered, created, keyid, keyid_raw, alg, signature, date_header = _parse_signature_headers(headers)
    now = int(time.time())
    if abs(now - created) > 300:
        raise AidError("ERR_SECURITY", "Signature created timestamp outside acceptance window")
    if date_header:
        try:
            from email.utils import parsedate_to_datetime  # stdlib

            dt = parsedate_to_datetime(date_header)
            epoch = int(dt.timestamp())
        except ValueError as e:
            logging.exception(f"Failed to parse Date header: {e}")
            raise AidError("ERR_SECURITY", "Invalid Date header") from None
        if abs(now - epoch) > 300:
            raise AidError("ERR_SECURITY", "HTTP Date header outside acceptance window")
    if not hmac.compare_digest(keyid.encode('utf-8'), kid.encode('utf-8')):
        raise AidError("ERR_SECURITY", "Signature keyid mismatch")
    if not hmac.compare_digest(alg.encode('utf-8'), b"ed25519"):
        raise AidError("ERR_SECURITY", "Unsupported signature algorithm")

    host = parsed.netloc
    base = _build_signature_base(
        covered,
        created=created,
        keyid=keyid_raw,
        alg=alg,
        method="GET",
        target_uri=uri,
        host=host,
        date=date_header or date_hdr,
        challenge=challenge,
    )
    if os.environ.get("AID_DEBUG_PKA") == "1":
        _debug_write("base_runtime.txt", base.decode("utf-8", errors="ignore"))

    pub = _multibase_decode(pka)
    if len(pub) != 32:
        raise AidError("ERR_SECURITY", "Invalid PKA length")

    try:
        # Prefer PyNaCl if available
        try:
            from nacl.signing import VerifyKey  # type: ignore

            vk = VerifyKey(pub)
            vk.verify(base, signature)  # raises on failure
            return
        except ImportError as e:
            logging.debug(f"PyNaCl not available, falling back to cryptography: {e}")
            # Fallback to cryptography if available
            from cryptography.hazmat.primitives.asymmetric import ed25519  # type: ignore
            from cryptography.exceptions import InvalidSignature  # type: ignore

            vk = ed25519.Ed25519PublicKey.from_public_bytes(pub)
            try:
                vk.verify(signature, base)
            except InvalidSignature:
                raise AidError("ERR_SECURITY", "PKA signature verification failed") from None
    except AidError:
        raise
    except Exception as exc:  # pragma: no cover - missing libs
        raise AidError("ERR_SECURITY", f"PKA verification unavailable: {exc}") from None
